File: bindings/fortran/src/safe_path.f90
! SPDX-License-Identifier: PMPL-1.0-or-later
! Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
!
! Proven SafePath - Directory traversal prevention for Fortran
!
! Thin wrapper around libproven's verified SafePath module.
! All path analysis is executed in the Idris 2 core via the Zig FFI
! bridge; this module only marshals Fortran strings to/from C pointers.
!
! Link with: -lproven

module safe_path
    use, intrinsic :: iso_c_binding
    use proven_ffi
    implicit none
    private

    !> Result type for path operations
    type, public :: PathResult
        character(len=:), allocatable :: path
        logical                      :: ok     = .false.
        integer(c_int32_t)           :: status = PROVEN_ERR_NOT_IMPLEMENTED
    end type PathResult

    public :: has_traversal, sanitize_filename

contains

    ! =========================================================================
    ! Check for directory traversal sequences  ->  proven_path_has_traversal
    ! =========================================================================
    function has_traversal(path) result(is_dangerous)
        character(len=*), intent(in) :: path
        logical :: is_dangerous
        character(len=len_trim(path), kind=c_char), target :: c_buf
        type(c_proven_bool_result) :: c_res
        integer :: trimmed_len

        trimmed_len = len_trim(path)
        c_buf = path(1:trimmed_len)

        c_res = proven_path_has_traversal(c_loc(c_buf), int(trimmed_len, c_size_t))

        if (c_res%status == PROVEN_OK) then
            is_dangerous = logical(c_res%value)
        else
            ! Conservative default: treat as dangerous on error
            is_dangerous = .true.
        end if
    end function has_traversal

    ! =========================================================================
    ! Sanitize filename  ->  proven_path_sanitize_filename
    ! =========================================================================
    function sanitize_filename(input) result(output)
        character(len=*), intent(in) :: input
        character(len=:), allocatable :: output
        character(len=len_trim(input), kind=c_char), target :: c_buf
        type(c_proven_string_result) :: c_res
        character(len=1, kind=c_char), pointer :: f_ptr(:)
        integer :: trimmed_len

        trimmed_len = len_trim(input)
        c_buf = input(1:trimmed_len)

        c_res = proven_path_sanitize_filename(c_loc(c_buf), int(trimmed_len, c_size_t))

        if (c_res%status == PROVEN_OK .and. c_associated(c_res%value)) then
            call c_f_pointer(c_res%value, f_ptr, [c_res%length])
            allocate(character(len=c_res%length) :: output)
            block
                integer :: i
                do i = 1, int(c_res%length)
                    output(i:i) = f_ptr(i)
                end do
            end block
            call proven_free_string(c_res%value)
        else
            output = input(1:trimmed_len)
        end if
    end function sanitize_filename

end module safe_path

File: bindings/rust/src/safe_path.rs
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>

//! Safe filesystem path operations via libproven FFI.
//!
//! Prevents directory traversal attacks and sanitizes filenames.
//! All operations delegate to Idris 2 verified code.

use crate::core::{self, Result};
use crate::ffi;

/// Safe path operations that prevent traversal attacks.
pub struct SafePath;

impl SafePath {
    /// Check if a path contains directory traversal sequences (e.g., `../`).
    ///
    /// Returns `true` if traversal is detected.
    pub fn has_traversal(path: &str) -> Result<bool> {
        let bytes = path.as_bytes();
        // SAFETY: We pass a valid pointer and length from the string's bytes.
        // The FFI function only reads from the pointer within the given length.
        let result = unsafe {
            ffi::proven_path_has_traversal(bytes.as_ptr(), bytes.len())
        };
        core::bool_result_to_result(result)
    }

    /// Sanitize a filename by removing dangerous characters.
    ///
    /// Returns the sanitized filename as a `String`.
    pub fn sanitize_filename(filename: &str) -> Result<String> {
        let bytes = filename.as_bytes();
        // SAFETY: We pass a valid pointer and length from the string's bytes.
        // The FFI function only reads from the pointer and allocates a new
        // string that we free via string_result_to_result.
        let result = unsafe {
            ffi::proven_path_sanitize_filename(bytes.as_ptr(), bytes.len())
        };
        core::string_result_to_result(result)
    }
}
